spring-integration-core/src/main/java/org/springframework/integration/endpoint/AbstractPollingEndpoint.java

```diff
@@ -343,11 +346,21 @@ public void run() {
 						count++;
 					}
 					catch (Exception e) {
-						if (e instanceof RuntimeException) {
-							throw (RuntimeException) e;
+						if (e instanceof MessagingException) {
+							throw (MessagingException) e;
 						}
 						else {
-							throw new MessageHandlingException(new ErrorMessage(e), e);
+							Message<?> failedMessage = null;
+							Object resource = TransactionSynchronizationManager.getResource(getResourceToBind());
+							if (resource instanceof IntegrationResourceHolder) {
+								failedMessage = ((IntegrationResourceHolder) resource).getMessage();
+							}
+							throw new MessagingException(failedMessage, e);
+						}
+					}
+					finally {
+						if (TransactionSynchronizationManager.hasResource(getResourceToBind())) {
+							TransactionSynchronizationManager.unbindResource(getResourceToBind());
```

Comment:
Minor, but we can extract local variable for the `getResourceToBind()`.
Not sure if compiler does an optimization.
Although it can't because of possibility to have a different result from each method invocation.